graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/TuplePrototypeBuiltins.java

```diff
@@ -0,0 +1,995 @@
+                toObjectNode = insert(JSToObjectNode.createToObjectNoCheck(getContext()));
+            }
+            return toObjectNode.execute(object);
+        }
+
+        private long getLengthOfArrayLike(Object obj) {
+            if (getLengthNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                getLengthNode = insert(JSGetLengthNode.create(getContext()));
+            }
+            return getLengthNode.executeLong(obj);
+        }
+
+        private boolean hasProperty(Object obj, long idx) {
+            if (hasPropertyNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                hasPropertyNode = insert(JSHasPropertyNode.create());
+            }
+            return hasPropertyNode.executeBoolean(obj, idx);
+        }
+
+        private Object get(Object target, long index) {
+            if (readElementNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                readElementNode = insert(ReadElementNode.create(getContext()));
+            }
+            if (JSRuntime.longIsRepresentableAsInt(index)) {
+                return readElementNode.executeWithTargetAndIndex(target, (int) index);
+            } else {
+                return readElementNode.executeWithTargetAndIndex(target, (double) index);
+            }
+        }
+    }
+
+    public abstract static class JSTupleFilterNode extends BasicTupleOperation {
+
+        private final BranchProfile growProfile = BranchProfile.create();
+
+        public JSTupleFilterNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        protected Tuple filter(Object thisObj, Object callbackfn, Object thisArg,
+                               @Cached IsCallableNode isCallableNode,
+                               @Cached JSToBooleanNode toBooleanNode,
+                               @Cached("createCall()") JSFunctionCallNode callNode) {
+            Tuple tuple = thisTupleValue(thisObj);
+            if (!isCallableNode.executeBoolean(callbackfn)) {
+                throw Errors.createTypeErrorCallableExpected();
+            }
+            SimpleArrayList<Object> list = SimpleArrayList.create(tuple.getArraySize());
+            for (long k = 0; k < tuple.getArraySize(); k++) {
+                Object value = tuple.getElement(k);
+                boolean selected = toBooleanNode.executeBoolean(
+                        callNode.executeCall(JSArguments.create(thisArg, callbackfn, value, k, tuple))
+                );
+                if (selected) {
+                    list.add(value, growProfile);
+                }
+            }
+            return Tuple.create(list.toArray());
+        }
+    }
+
+    public abstract static class TupleFlattenOperation extends BasicTupleOperation {
+
+        private final BranchProfile errorProfile = BranchProfile.create();
```

Comment:
du hast in `BasicTupleOperation` schon ein errorProfile. Setz es dort auf `protected` und nutze es auch hier, brauchst nicht ein neues anlegen.

Reply:
Done